File: mydia-rs/crates/indexers/src/search_result.rs
//! Normalized search-result struct.
//!
//! Port of `lib/mydia/indexers/search_result.ex` plus the
//! `Mydia.Indexers.Structs.{QualityInfo, SearchResultMetadata, RankedResult, ScoreBreakdown}`
//! companions. Every adapter normalizes its responses into [`SearchResult`].

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

/// Download protocol discriminant for a result. `:torrent` and `:nzb`
/// in Elixir.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum DownloadProtocol {
    Torrent,
    Nzb,
}

/// Parsed quality information lifted out of a release title.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct QualityInfo {
    pub resolution: Option<String>,
    pub source: Option<String>,
    pub codec: Option<String>,
    pub audio: Option<String>,
    #[serde(default)]
    pub hdr: bool,
    pub hdr_format: Option<String>,
    #[serde(default)]
    pub proper: bool,
    #[serde(default)]
    pub repack: bool,
}

impl QualityInfo {
    /// Returns true when every field is unset (apart from boolean flags).
    #[must_use]
    pub fn is_empty(&self) -> bool {
        self.resolution.is_none()
            && self.source.is_none()
            && self.codec.is_none()
            && self.audio.is_none()
    }

    /// Render a compact human label — `"1080p BluRay x264 DTS"`.
    #[must_use]
    pub fn format(&self) -> String {
        let mut parts: Vec<String> = Vec::new();
        if let Some(r) = &self.resolution {
            parts.push(r.clone());
        }
        if let Some(s) = &self.source {
            parts.push(s.clone());
        }
        if let Some(c) = &self.codec {
            parts.push(c.clone());
        }
        if let Some(a) = &self.audio {
            parts.push(a.clone());
        }
        if self.hdr {
            if let Some(fmt) = &self.hdr_format {
                parts.push(fmt.clone());
            } else {
                parts.push("HDR".into());
            }
        }
        if self.proper {
            parts.push("PROPER".into());
        }
        if self.repack {
            parts.push("REPACK".into());
        }
        parts.join(" ")
    }
}

/// Extra metadata typically attached to season-pack results.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct SearchResultMetadata {
    #[serde(default)]
    pub season_pack: bool,
    pub season_number: Option<i32>,
    pub episode_count: Option<i32>,
    #[serde(default)]
    pub episode_ids: Vec<String>,
}

impl SearchResultMetadata {
    #[must_use]
    pub fn season_pack(season_number: i32, episode_count: i32, episode_ids: Vec<String>) -> Self {
        Self {
            season_pack: true,
            season_number: Some(season_number),
            episode_count: Some(episode_count),
            episode_ids,
        }
    }
}

/// Normalized search result. All adapters return this shape.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SearchResult {
    pub title: String,
    pub size: u64,
    pub seeders: Option<u32>,
    pub leechers: Option<u32>,
    pub download_url: String,
    pub info_url: Option<String>,
    pub indexer: String,
    pub category: Option<i32>,
    pub published_at: Option<DateTime<Utc>>,
    pub quality: Option<QualityInfo>,
    pub metadata: Option<SearchResultMetadata>,
    pub tmdb_id: Option<i64>,
    pub tvdb_id: Option<i64>,
    pub imdb_id: Option<String>,
    pub download_protocol: Option<DownloadProtocol>,
    pub usenet_date: Option<DateTime<Utc>>,
    pub nzb_completion: Option<f64>,
    pub nzb_grabs: Option<u32>,
    pub guid: Option<String>,
}

impl SearchResult {
    /// Healthy seeder→leecher ratio normalized to `[0.0, 1.0]`.
    #[must_use]
    pub fn health_score(&self) -> f64 {
        let Some(seeders) = self.seeders else {
            return 0.0;
        };
        let leechers = self.leechers.unwrap_or(0);
        let total = seeders + leechers;
        if total == 0 {
            return 0.0;
        }
        if seeders == 0 {
            return 0.1;
        }
        let denom = f64::from(seeders) + f64::from(leechers);
        (f64::from(seeders) / denom + f64::from(seeders) / 100.0).min(1.0)
    }

    /// Render the size as a human-friendly string (`"1.0 GB"`).
    #[must_use]
    pub fn format_size(&self) -> String {
        format_bytes(self.size)
    }

    /// Concise quality description, falls back to `"Unknown"`.
    #[must_use]
    pub fn quality_description(&self) -> String {
        match &self.quality {
            None => "Unknown".into(),
            Some(q) => {
                let formatted = q.format();
                if formatted.is_empty() {
                    "Unknown".into()
                } else {
                    formatted
                }
            }
        }
    }
}

/// Builder for a [`SearchResult`]. Mirrors the Elixir `SearchResult.new/1`
/// keyword API while leaving fields strongly typed.
#[derive(Debug, Default)]
pub struct SearchResultBuilder {
    pub title: Option<String>,
    pub size: Option<u64>,
    pub seeders: Option<u32>,
    pub leechers: Option<u32>,
    pub download_url: Option<String>,
    pub info_url: Option<String>,
    pub indexer: Option<String>,
    pub category: Option<i32>,
    pub published_at: Option<DateTime<Utc>>,
    pub quality: Option<QualityInfo>,
    pub metadata: Option<SearchResultMetadata>,
    pub tmdb_id: Option<i64>,
    pub tvdb_id: Option<i64>,
    pub imdb_id: Option<String>,
    pub download_protocol: Option<DownloadProtocol>,
    pub usenet_date: Option<DateTime<Utc>>,
    pub nzb_completion: Option<f64>,
    pub nzb_grabs: Option<u32>,
    pub guid: Option<String>,
}

impl SearchResultBuilder {
    pub fn new() -> Self {
        Self::default()
    }

    /// Build, returning `None` if any of the enforced fields are missing
    /// (title, size, seeders, leechers, `download_url`, indexer in the
    /// Elixir source — we surface this as a `Result` so callers can route
    /// the error properly).
    pub fn build(self) -> Result<SearchResult, &'static str> {
        Ok(SearchResult {
            title: self.title.ok_or("title is required")?,
            size: self.size.ok_or("size is required")?,
            seeders: self.seeders,
            leechers: self.leechers,
            download_url: self.download_url.ok_or("download_url is required")?,
            info_url: self.info_url,
            indexer: self.indexer.ok_or("indexer is required")?,
            category: self.category,
            published_at: self.published_at,
            quality: self.quality,
            metadata: self.metadata,
            tmdb_id: self.tmdb_id,
            tvdb_id: self.tvdb_id,
            imdb_id: self.imdb_id,
            download_protocol: self.download_protocol,
            usenet_date: self.usenet_date,
            nzb_completion: self.nzb_completion,
            nzb_grabs: self.nzb_grabs,
            guid: self.guid,
        })
    }
}

fn format_bytes(bytes: u64) -> String {
    const KB: f64 = 1024.0;
    const MB: f64 = 1024.0 * 1024.0;
    const GB: f64 = 1024.0 * 1024.0 * 1024.0;
    let b = bytes as f64;
    if b < KB {
        format!("{bytes} B")
    } else if b < MB {
        format!("{:.1} KB", b / KB)
    } else if b < GB {
        format!("{:.1} MB", b / MB)
    } else {
        format!("{:.1} GB", b / GB)
    }
}

/// Score breakdown for a ranked result.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScoreBreakdown {
    pub quality: f64,
    pub seeders: f64,
    pub size: f64,
    pub age: f64,
    pub title_match: f64,
    pub tag_bonus: f64,
    pub total: f64,
}

/// A [`SearchResult`] decorated with its ranking score.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RankedResult {
    pub result: SearchResult,
    pub score: f64,
    pub breakdown: ScoreBreakdown,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn result(seeders: Option<u32>, leechers: Option<u32>) -> SearchResult {
        SearchResultBuilder {
            title: Some("Ubuntu 22.04".into()),
            size: Some(1_073_741_824),
            seeders,
            leechers,
            download_url: Some("magnet:?xt=urn:btih:abc".into()),
            indexer: Some("Prowlarr".into()),
            ..SearchResultBuilder::new()
        }
        .build()
        .unwrap()
    }

    #[test]
    fn health_score_returns_zero_for_nil_seeders() {
        let r = result(None, None);
        assert!((r.health_score() - 0.0).abs() < f64::EPSILON);
    }

    #[test]
    fn health_score_caps_at_one() {
        let r = result(Some(10_000), Some(0));
        assert!((r.health_score() - 1.0).abs() < f64::EPSILON);
    }

    #[test]
    fn format_size_renders_gigabytes() {
        let r = result(Some(10), Some(5));
        assert_eq!(r.format_size(), "1.0 GB");
    }

    #[test]
    fn quality_description_falls_back_to_unknown() {
        let r = result(Some(10), Some(5));
        assert_eq!(r.quality_description(), "Unknown");
    }

    #[test]
    fn quality_format_includes_all_set_parts() {
        let q = QualityInfo {
            resolution: Some("1080p".into()),
            source: Some("BluRay".into()),
            codec: Some("x264".into()),
            audio: Some("DTS".into()),
            hdr: true,
            hdr_format: Some("DV".into()),
            proper: true,
            repack: false,
        };
        assert_eq!(q.format(), "1080p BluRay x264 DTS DV PROPER");
    }

    #[test]
    fn season_pack_helper_sets_flag() {
        let m = SearchResultMetadata::season_pack(1, 10, vec!["a".into(), "b".into()]);
        assert!(m.season_pack);
        assert_eq!(m.season_number, Some(1));
        assert_eq!(m.episode_count, Some(10));
        assert_eq!(m.episode_ids.len(), 2);
    }
}

File: lib/mydia/indexers/search_result.ex
defmodule Mydia.Indexers.SearchResult do
  @moduledoc """
  Normalized search result structure.

  This struct represents a torrent search result from any indexer in a
  normalized format. All adapters should convert their responses to this
  common structure.

  ## Fields

    * `:title` - Release title/name
    * `:size` - File size in bytes
    * `:seeders` - Number of seeders
    * `:leechers` - Number of leechers (peers)
    * `:download_url` - Magnet link or torrent file URL
    * `:info_url` - Link to more information about the release (optional)
    * `:indexer` - Name of the indexer that returned this result
    * `:category` - Category ID from the indexer
    * `:published_at` - When the torrent was published (optional)
    * `:quality` - Parsed quality information (resolution, codec, etc.)
    * `:metadata` - Additional metadata (e.g., season pack info) (optional)
    * `:tmdb_id` - TMDB ID from indexer (optional, for ID-based matching)
    * `:imdb_id` - IMDB ID from indexer (optional, for ID-based matching)
    * `:usenet_date` - When the release was originally posted to Usenet (optional, NZB only)
    * `:nzb_completion` - Article completion ratio (0.0..1.0) reported by Usenet indexer (optional)
    * `:nzb_grabs` - Number of times this release has been grabbed (NZB popularity indicator)
    * `:guid` - Indexer-provided stable identifier for the release (optional).
      Used by the release blacklist (issue #123) to match repeat failures. When
      missing, the blacklist falls back to a SHA-256 of (indexer, title, size).

  ## Quality Information

  The `:quality` field contains parsed quality information extracted from
  the release title as a `Mydia.Library.Structs.Quality` struct:

      %Quality{
        resolution: "1080p" | "720p" | "2160p" | "480p" | nil,
        source: "BluRay" | "WEB-DL" | "WEBRip" | "HDTV" | nil,
        codec: "x264" | "x265" | "H.264" | "H.265" | nil,
        audio: "AAC" | "AC3" | "DTS" | "TrueHD" | nil,
        hdr_format: :hdr10 | :hdr10_plus | :hlg | nil,
        dolby_vision: true | false,
        proper: true | false,
        repack: true | false
      }

  ## Examples

      iex> %SearchResult{
      ...>   title: "Ubuntu 22.04 LTS 1080p BluRay x264",
      ...>   size: 4_294_967_296,
      ...>   seeders: 100,
      ...>   leechers: 50,
      ...>   download_url: "magnet:?xt=urn:btih:...",
      ...>   indexer: "Prowlarr",
      ...>   category: 2000,
      ...>   quality: %{resolution: "1080p", source: "BluRay", codec: "x264"}
      ...> }
  """

  alias Mydia.Library.Structs.Quality
  alias Mydia.Indexers.Structs.SearchResultMetadata

  @type t :: %__MODULE__{
          title: String.t(),
          size: non_neg_integer(),
          seeders: non_neg_integer() | nil,
          leechers: non_neg_integer() | nil,
          download_url: String.t(),
          info_url: String.t() | nil,
          indexer: String.t(),
          category: integer() | nil,
          published_at: DateTime.t() | nil,
          quality: Quality.t() | nil,
          metadata: SearchResultMetadata.t() | nil,
          tmdb_id: integer() | nil,
          tvdb_id: integer() | nil,
          imdb_id: String.t() | nil,
          download_protocol: :torrent | :nzb | nil,
          usenet_date: DateTime.t() | nil,
          nzb_completion: float() | nil,
          nzb_grabs: non_neg_integer() | nil,
          guid: String.t() | nil
        }

  @enforce_keys [:title, :size, :seeders, :leechers, :download_url, :indexer]
  defstruct [
    :title,
    :size,
    :seeders,
    :leechers,
    :download_url,
    :info_url,
    :indexer,
    :category,
    :published_at,
    :quality,
    :metadata,
    :tmdb_id,
    :tvdb_id,
    :imdb_id,
    :download_protocol,
    :usenet_date,
    :nzb_completion,
    :nzb_grabs,
    :guid
  ]

  @info_url_schemes ~w(http https)

  @doc """
  Creates a new search result with default values.

  ## Examples

      iex> SearchResult.new(
      ...>   title: "Ubuntu 22.04",
      ...>   size: 1_000_000_000,
      ...>   seeders: 50,
      ...>   leechers: 10,
      ...>   download_url: "magnet:?xt=...",
      ...>   indexer: "Prowlarr"
      ...> )
      %SearchResult{...}
  """
  @spec new(keyword()) :: t()
  def new(attrs) do
    struct!(__MODULE__, attrs)
  end

  @doc """
  Calculates the health score of a torrent based on seeders and leechers.

  Returns a value between 0.0 and 1.0, where higher is better.

  ## Examples

      iex> result = %SearchResult{seeders: 100, leechers: 50}
      iex> SearchResult.health_score(result)
      0.95

      iex> result = %SearchResult{seeders: 0, leechers: 0}
      iex> SearchResult.health_score(result)
      0.0
  """
  @spec health_score(t()) :: float()
  def health_score(%__MODULE__{seeders: nil}), do: 0.0

  def health_score(%__MODULE__{seeders: seeders, leechers: leechers}) do
    leechers = leechers || 0
    total = seeders + leechers

    cond do
      total == 0 -> 0.0
      seeders == 0 -> 0.1
      true -> min(1.0, seeders / (seeders + leechers) + seeders / 100)
    end
  end

  @doc """
  Formats the file size in a human-readable format.

  ## Examples

      iex> result = %SearchResult{size: 1_073_741_824}
      iex> SearchResult.format_size(result)
      "1.0 GB"

      iex> result = %SearchResult{size: 1_048_576}
      iex> SearchResult.format_size(result)
      "1.0 MB"
  """
  @spec format_size(t()) :: String.t()
  def format_size(%__MODULE__{size: size}) do
    format_bytes(size)
  end

  @doc """
  Returns a description of the quality for display.

  ## Examples

      iex> result = %SearchResult{quality: %{resolution: "1080p", source: "BluRay", codec: "x264"}}
      iex> SearchResult.quality_description(result)
      "1080p BluRay x264"

      iex> result = %SearchResult{quality: nil}
      iex> SearchResult.quality_description(result)
      "Unknown"
  """
  @spec quality_description(t()) :: String.t()
  def quality_description(%__MODULE__{quality: nil}), do: "Unknown"

  def quality_description(%__MODULE__{quality: quality}) do
    # Build a concise quality badge including all quality attributes
    parts =
      [
        quality.resolution,
        quality.source,
        quality.codec,
        quality.audio,
        Quality.hdr?(quality) && "HDR",
        quality.proper && "PROPER",
        quality.repack && "REPACK"
      ]
      |> Enum.filter(& &1)

    case Enum.join(parts, " ") do
      "" -> "Unknown"
      description -> description
    end
  end

  @doc """
  Returns a URL safe to render as a link to the release's page on the indexer.

  Returns `nil` unless `:info_url` is an absolute `http`/`https` URL with a host
  that differs from `:download_url`.

  Indexers supply this value, and Cardigann definitions are fetched from an
  external repository, so the scheme allowlist is a security boundary: HEEx does
  not validate URL schemes, and a `javascript:` string would otherwise render as
  a live script URL. The `:download_url` check exists because Prowlarr and
  Jackett fall back to the torznab `<guid>`, which on some indexers is the
  `.torrent` link rather than a details page.

  ## Examples

      iex> result = %SearchResult{title: "T", size: 1, seeders: 1, leechers: 1,
      ...>   download_url: "magnet:?xt=urn:btih:abc", indexer: "X",
      ...>   info_url: "https://tracker.example/details/42"}
      iex> SearchResult.info_page_url(result)
      "https://tracker.example/details/42"

      iex> result = %SearchResult{title: "T", size: 1, seeders: 1, leechers: 1,
      ...>   download_url: "magnet:?xt=urn:btih:abc", indexer: "X",
      ...>   info_url: "javascript:alert(1)"}
      iex> SearchResult.info_page_url(result)
      nil
  """
  @spec info_page_url(t()) :: String.t() | nil
  def info_page_url(%__MODULE__{info_url: info_url, download_url: download_url})
      when is_binary(info_url) do
    trimmed = String.trim(info_url)

    cond do
      trimmed == "" -> nil
      trimmed == download_url -> nil
      not linkable_scheme?(trimmed) -> nil
      true -> trimmed
    end
  end

  def info_page_url(%__MODULE__{}), do: nil

  # Private helpers

  defp linkable_scheme?(url) do
    case URI.parse(url) do
      %URI{scheme: scheme, host: host} when scheme in @info_url_schemes ->
        is_binary(host) and host != ""

      _ ->
        false
    end
  end

  defp format_bytes(bytes) when bytes < 1024, do: "#{bytes} B"

  defp format_bytes(bytes) when bytes < 1024 * 1024 do
    kb = bytes / 1024
    "#{Float.round(kb, 1)} KB"
  end

  defp format_bytes(bytes) when bytes < 1024 * 1024 * 1024 do
    mb = bytes / (1024 * 1024)
    "#{Float.round(mb, 1)} MB"
  end

  defp format_bytes(bytes) do
    gb = bytes / (1024 * 1024 * 1024)
    "#{Float.round(gb, 1)} GB"
  end
end
